### ISSUE TITLE
review assertions and avoid side effects

### ISSUE THREAD

**Issue by hfp** [cp2k/cp2k#922]
Input files `benchmarks/QS_mp2_rpa/*/*.inp` stopped working and cause program abort with:

```text
 *******************************************************************************
 *   ___                                                                       *
 *  /   \                                                                      *
 * [ABORT]                                                                     *
 *  \___/    The requested basis set <RI_AUX> for element <O> was not found in *
 *    |                      the basis set files <./BASIS_H2O>                 *
 *  O/|                                                                        *
 * /| |                                                                        *
 * / \                                          aobasis/basis_set_types.F:1540 *
 *******************************************************************************


 ===== Routine Calling Stack =====

            3 read_qs_kind
            2 create_qs_kind_set
            1 CP2K
```

The complained basis set seems to be present in `BASIS_H2O` ("`O    RI-cc-QZ`").

**Comment by hfp** [comment]
This is caused by a miscompiled binary. I was using GCC 8.3 and 9.3 with quite some optimization flags. I will check which flag(s) is(are) causing the issue.

**Comment by hfp** [comment]
This is caused by `NDEBUG` applied to CP2K´s source code (which happened accidentally on my end). The problem with `NDEBUG` was discovered soon after its introduction but remained unspecific. However, this issue confirms/narrows a potential problem with eliding assertions (side effects). Side note: DBCSR´s source code is fine with eliding assertions (it is currently hard to apply `NDEBUG` separately to CP2K and DBCSR).

**Comment by oschuett** [comment]
Yes, it's unfortunate that some of our assertions have side-effects. Since there is no automatic way to detect side-effects, should we maybe just execute `CPASSERT` regardless of `NDEBUG`?

**Comment by oschuett** [comment]
It's certainly good to once review all assertions. However, unless we find an automatic way to detect side-effects, they will reappear. So, I think unconditional execution is our safest option.

>  Was the latter considered or is calling exit the way to go?

The `EXIT` statement is never actually executed, because we call [cp_abort_handler](https://github.com/cp2k/cp2k/blob/master/src/common/cp_error_handling.F#L54) before. The `EXIT` was merely meant as a compiler hint (not sure any compiler every picked up on it).

I'm not familiar with breakpoints. Would we still get the same error message and stack trace when no debugger is attached?

**Comment by hfp** [comment]
Sorry, I meant the C-exit (not F´s `EXIT`).  Ok, this tells me I can give it a try and test this wrt stacktrace, i.e., if no debugger is attached it should call exit (and thereby run the registered exit-handler and present the stacktrace).

> So, I think unconditional execution is our safest option.

I think we have several tests failing if CP2K is built with `NDEBUG`. It is therefore save to keep the normal behavior (especially if it is not used/enabled by default). Let me also double-check that tests are failing and report the case here.

Regarding automatic detection,  I cannot see a viable way to identify side-effects automatically based on static analysis. Running the tests is the best option.

**Comment by oschuett** [comment]
> I think we have several tests failing if CP2K is built with `NDEBUG`.

That's certainly comforting. However, currently none of our dashboard tests is using `NDEBUG`. And enabling it could hide other problems that `CPASSERT` would catch. So, I'm wondering what's the upside of supporting `NDEBUG`? Did you find any assertions in tight loops? That would be a misuse of `CPASSERT`, which we should refactor.

**Comment by hfp** [comment]
Fixed by #957: `-DNDEBUG` can now be applied to the entire code base (CP2K and DBCSR) and all regtests are passing. Perhaps we want to implement a "release build" which defines `-DNDEBUG`.

**Comment by oschuett** [comment]
> Perhaps we want to implement a "release build" which defines -DNDEBUG.

While `CPASSERT` should not be used for user facing errors, they still catch a lot of corner case and unsupported feature combinations. Hence, I really think one should always have asserts enabled.

**Comment by hfp** [comment]
I am regularly testing CP2K/master with `-DNDEBUG` and there was no such issue recently (or since the fixes under #922).
